### PR TITLE
Removed `evaluate=False` from the Lark-based LaTeX parser

### PR DESCRIPTION
**Pull request by wermos**
#### References to other Issues or PRs

Addresses #24116 for the Lark-based LaTeX parser.

#### Brief description of what is fixed or changed

Originally, the ANTLR LaTeX parser returned unevaluated expressions. There were concerns that unevaluated expressions can [create all sorts of issues ](https://github.com/sympy/sympy/issues/24116#issue-1398475575), so we wanted to change this behavior.

I made it so that both `evaluate=False` and `evaluate=True` work with the Lark-based LaTeX parser, with `evaluate=True` being the default.

#### Release Notes

* parsing
    * The Lark-based LaTeX parser now evaluates the input string and returns the result. For example, `"\sqrt{\frac{12}{6}}"` will re turn `\sqrt(2)` now. The previous behavior can achieved by using the `evaluate(False)` context manager.

### CODE REVIEW

**Comment by wermos** [comment]
cc @sylee957 and @Upabjojr for review

File: sympy/parsing/latex/lark/latex_parser.py [hunk 1]
```diff
@@ -114,33 +114,33 @@ def relation(self, tokens):
...
-        return sympy.Mul(tokens[0], sympy.Pow(tokens[2], -1, evaluate=False), evaluate=False)
+        return sympy.Mul(tokens[0], sympy.Pow(tokens[2], -1))
 
     def superscript(self, tokens):
-        return sympy.Pow(tokens[0], tokens[2], evaluate=False)
+        return sympy.Pow(tokens[0], tokens[2])
 
     def fraction(self, tokens):
-        return sympy.Mul(tokens[1], sympy.Pow(tokens[2], -1, evaluate=False), evaluate=False)
+        return sympy.Mul(tokens[1], sympy.Pow(tokens[2], -1))
```
Comment (sylee957) [inline sympy/parsing/latex/lark/latex_parser.py:140]: After removing `evaluate` option, it is now possible to build sympy expressions without `Add, Mul, Pow`. I'd prefer to write division as `tokens[1] / tokens[2]`

Comment (wermos) [reply sympy/parsing/latex/lark/latex_parser.py:140]: How would this change interact with the `evaluate(False)` context manager? I.e., would
```python
with evaluate(False):
    parse_latex_lark(r"\frac12")
```
still work as expected?

Comment (wermos) [reply sympy/parsing/latex/lark/latex_parser.py:140]: It looks like it works as expected:
```python
In [14]: with evaluate(False):
    ...:    expr = parse_latex_lark(r"\frac12")
In [15]: expr
Out[15]: 1/2
In [16]: srepr(expr)
Out[16]: 'Mul(Integer(1), Pow(Integer(2), Integer(-1)))'
```

Comment (sylee957) [reply sympy/parsing/latex/lark/latex_parser.py:140]: Using `evaluate(False)` context manager means that `evaluate=False` is passed throughout any depth inside the block,
without drilling down `evaluate=False`, so it works.

Comment (wermos) [reply sympy/parsing/latex/lark/latex_parser.py]: I've made the requested change, so I'm marking this conversation as resolved.

**Comment by wermos** [comment]
@sylee957 Some tests for the `evaluate=False` portion are failing. Could it be due to the changes I made like converting `sympy.Add` to `+` and `sympy.Mul` to `*` and so on?

**Comment by wermos** [comment]
@sylee957 I've reverted the change that you suggested because tests were failing due to it. I'm not convinced that we should remove support for `evaluate=False` yet, so it's important that we support both for now.

**Comment by sylee957** [comment]
The reason that we don’t have to support that is because evaluate(false) context manager can support that behavior orthogonally.

I don’t think that we should really add any options for sake of the future. And we are also very unsure supporting evaluate=False in the future as well, because that doesn’t work with lots of other sympy features, so I don’t want to make the surface area more larger.

**Comment by wermos** [comment]
> The reason that we don’t have to support that is because evaluate(false) context manager can support that behavior orthogonally.

When I removed the `sympy.Add` constructor call and added `+` to the transformer for the `add` node, the tests were failing. Things like "1 + 1" were being parsed as "2" even with `evaluate=False`. So I'm not sure that the `evaluate(False)` context manager is working as intended in those cases. I'm not sure if this is a bug in my code or something to do with the context manager.

> And we are also very unsure supporting evaluate=False in the future as well, because that doesn’t work with lots of other sympy features, so I don’t want to make the surface area more larger for maintanance.

That's a good point. While I personally have no issue with making the change, I'd feel better if we got Aaron and Francesco's thoughts on this. Perhaps we can discuss the matter on the Matrix chat. I'll bring it up there, and also cc them here.

cc @asmeurer @Upabjojr for thoughts: Do we want to support `evaluate=False` at all?

**Comment by sylee957** [comment]
I also recognize that some of the code path is not consistently returning SymPy objects

https://github.com/sympy/sympy/blob/5b1ad1dc047e3901d4e389c91ce9e95d45f4d166/sympy/parsing/latex/lark/latex_parser.py#L26
https://github.com/sympy/sympy/blob/5b1ad1dc047e3901d4e389c91ce9e95d45f4d166/sympy/parsing/latex/lark/latex_parser.py#L82

```
>>> from sympy import *
>>> from sympy.parsing.latex.lark import parse_latex_lark
>>> parse_latex_lark('1').__class__
<class 'int'>
```

My opinion is that these were the actual reason why `x + y` and `Add(x, y)` behave differently, and I also suggest to change that to return sympy `Integer` consistently, such that SymPy parser always returns sympy object anyway, and users don't get some surprise.

**Comment by oscarbenjamin** [comment]
> The reason that we don’t have to support that is because evaluate(false) context manager can support that behavior orthogonally.

I think that `evaluate(false)` is something that should be discouraged except when used around a block that does nothing more than create a simple expression so something like
```python
with evaluate(False):
      expr = parse_latex(text)
```
is probably a bad idea. The `evaluate(False)` might break all kinds of things if any part of the `parse_latex` implementation tries to create expressions as part of computing anything.

Writing any nontrivial code for the internals of SymPy that could reasonably be expected to work under the impact of `evaluate(False)` is not easy.

**Comment by oscarbenjamin** [comment]
> Passing the keyword `evaluate=False` everywhere has same effect

It is not the same effect because using `evaluate(False)` applies globally to all expressions that are created e.g.:
```python
In [5]: cat t.py
def func(a, b, evaluate=None):
    stuff1 = a + b
    stuff2 = Add(a, b, evaluate=evaluate)
    return stuff1, stuff2

In [8]: func(x, x)
Out[8]: (2⋅x, 2⋅x)

In [9]: func(x, x, evaluate=False)
Out[9]: (2⋅x, x + x)

In [10]: with evaluate(False):
    ...:     y = func(x, x)
    ...: 

In [11]: y
Out[11]: (x + x, x + x)
```
I agree that there are problems with unevaluated expressions but I think that `evaluate(False)` is much worse than passing `evaluate=False` explicitly because it can break any code that does not expect it from outside. I think that the use of `evaluate(False)` should definitely not be encouraged. If we want to support non-evaluation then passing `evaluate=False` should be the way to do it.

**Comment by sylee957** [comment]
I ask that high level functions of SymPy shouldn't that SymPy needs `evaluate=False` option at all, 
or it is better to make the SymPy classes don't evaluate by default, or have same interface of constructing the objects,
than implementing options outside that supports the behavior.

It already looks like having that as user option, is very unsatisfactory because the behavior it is complicated if evaluate=False is used shallowly or deeply.
and I'd rather introduce users with DIY approach, than using options of SymPy functions that skeptically implements that.

**Comment by sylee957** [comment]
And anyway I think that your example is artificial though
Semantically, if I see code like that, I rather criticize the code about what the *** are you using `+` than using `Add(a, b)` if you support `evaluate=` option.

And in fact, this shows the reason that having options about `evaluate` in SymPy high level functions, is very prone to mistakes, and we shouldn't do it.

```
def func(a, b, evaluate=None):
    stuff1 = a + b
    stuff2 = Add(a, b, evaluate=evaluate)
    return stuff1, stuff2
```

**Comment by sylee957** [comment]
The major problem about this is the roadmap problem, than its implementation detail.
And we really don't need to support legacy options like `evaluate=False` at all because we are building something new.

The real problem about roadmap is that we are not sure about we are not sure about how SymPy is going to handle eager evaluation, 
And we have no real analysis about if that option is wanted from users, or that is worth the cost.
You should drop `The previous behavior can achieved by using the evaluate(False) context manager.` from the release notes if that also doesn't need to be announced.

**Comment by oscarbenjamin** [comment]
I think that we should take this discussion to another issue that is specifically for discussing what to do about `evaluate=False` etc.

I do want to respond to this though:

> And we have no real analysis about if that option is wanted from users

It is very clear that many users want the option to prevent automatic evaluation somehow and that they would like SymPy to be able to make that work without everything just breaking. Especially there are many users who want this in the context of parsing. Those users really want to be able to parse an expression and get something that looks like an exact representation of what was parsed and that can then be printed back out in such a way that still looks exactly the same so e.g. (modulo whitespace, and braces etc) they want the following invariant:
```python
>>> latex(parse_latex(string)) == string
True
```

**Comment by wermos** [comment]
> I think that we should take this discussion to another issue that is specifically for discussing what to do about evaluate=False etc.

Just thought I'd chime in: The relevant issue for that is #24116.

**Comment by sylee957** [comment]
My opinion is that ‘we start minimal’
Adding things that seems trivial can be done anytime, but removing them because that doesn’t look good is more difficult.
I don’t want to make the discussion more longer because

However, as far as I have done during previous job, which actually had involved parsing SymPy expressions unevaluated.

I’ve just done by referencing the Transformer and overriding the rules individually,
Users know how to code, so we don’t have to supply everything. In fact, just a thin wrapper of transformer and grammar is enough for it, and just using evaluate=False was somewhat limited for me.

**Comment by wermos** [comment]
> they want the following invariant:
> ```python
>>> latex(parse_latex(string)) == string
> ```

While I got the sentiment, I'd like to point out that, as it's stated, this is not a meaningful metric:
```python
>>> string = r"\int\limits_0^1 \sin x dx"
>>> sympy_repr = parse_latex(string)
>>> print(sympy_repr)
Integral(sin(x), (x, 0, 1))
>>> latex_string = latex(sympy_repr)
>>> print(latex_string)
\int\limits_{0}^{1} \sin{\left(x \right)}\, dx
>>> latex_string == string
False

**Comment by sylee957** [comment]
I may request to fix the inconsistency issue of integers here

https://github.com/sympy/sympy/pull/25515#issuecomment-1678708129

Try to change int to Integer. We should always return SymPy objects anyway.
Python numbers behave differently in division.

**Comment by wermos** [comment]
> I may request to fix the inconsistency issue of integers here

Alright, I'll open a separate PR for that. I was originally planning to do that as a part of this PR, but it seems that we have a few blocking discussions that we need to resolve before this PR can find its way into `master`.

**Comment by sylee957** [comment]
I don’t think that this PR is blocking though. We agree to drop the evaluate=False as default, and we don’t have to add options here.

**Comment by wermos** [comment]
> I don’t think that this PR is blocking though. We agree to drop the evaluate=False as default, and we don’t have to add options here.

And what about `sympy.Add` vs `+`, etc.?

**Comment by sylee957** [comment]
I think that you need to fix the issues about integer returns before trying that. 
i think that there was a gain from testing out the arithmetic operator though, it revealed such hidden issue

**Comment by asmeurer** [comment]
Should the issue be closed?

**Comment by asmeurer** [comment]
Passing evaluate=False through everywhere does suck, because it's not the default so it has to manually be disabled everywhere. It would be much better if the classes just didn't evaluate, so the parser could just use Add(...), Mul(...) and so on, then have a single call to doit() at the end to evaluate everything. This is how things work in the matrix expressions, but not anywhere else. Changing this design in the core is not easy, though. The best we have right now is the evaluate(False) context manager which basically forces the core to work like this. Of course, we could have also fixed this by keeping the evaluate=False everywhere, but just evaluating by default at the end, and having a flag to disable it. But this fix ends up being cleaner and you can still achieve the same result with evaluate(False).

At any rate, I don't see an issue with `with evaluate(False): parse_latex(...)` as a supported pattern. evaluate(False) breaks things that try to do symbolic computation, but the latex parser shouldn't be doing that, and if it does it's probably a design mistake. It's probably worth at least having some tests for this. In particular, we do need to make sure that we are constructing expressions that would exactly correspond to the input when evaluate(False) is used, and that we aren't doing any symbolic math inside of the parser.

**Pull request merged**